React

ryanzigler opened the pull request:
@kongaiwen @MrReeko @BenChasin 
- Completed all api requests and db queries 
- All data (other than maps) is being properly imported and rendered on req
- Map renders, but does not use proper lat/long

ryanzigler committed:
- Created Headline component for hostel name, address, and icons
- Skeleton React index.jsx
- Updated/changed a couple of dependencies
- Webpack config is complete

ryanzigler committed:
- Completed all api requests and db queries
- All data (other than maps) is being properly imported on req
- Map renders, but does not use proper lat/long

ryanzigler committed:
fixed issue with npm run script and webpack

ryanzigler commented:
Screenshot of front-end as currently written:

![Screen Shot 2020-08-12 at 11 33 52 AM](https://user-images.githubusercontent.com/66638608/90042042-c31b4600-dc8f-11ea-9870-6d619f4ea042.png)

BenChasin reviewed (approved):
Looks great Ryan! Nice separation of concerns, everything was easy to read und understand.

ryancmoniz commented on client/index.jsx:
```diff
@@ -0,0 +1,40 @@
+        this.setState({ hostel: res.body });
+        this.componentDidMount();
+      })
+      .catch((err) => err);
```
A few things of note. You shouldn't directly call lifecycle methods in React. You can modify what they do when something happens (such as getting hostels when the component mounts). But if you feel the need to call it, you should probably modify the way your functions are working. Second, if we receive an error, what are we doing? Should we notify the user that something went wrong?

ryanzigler replied:
>You shouldn't directly call lifecycle methods in React. You can modify what they do when something happens (such as getting hostels when the component mounts). But if you feel the need to call it, you should probably modify the way your functions are working. 

I don't remember exactly what I was attempting to do and/or test with that at the time, but it was essentially a quick "fix" to test something. I has since been fixed and lifecycle methods are not being called directly any more.

> Second, if we receive an error, what are we doing? Should we notify the user that something went wrong?

I know that for certain errors, it makes sense to notify the user which is why we see 403 and 404 error pages on websites, but for projects like this where it doesn't make sense to build out an error page, is it a better idea to fail silently or to use something like `alert('blah, blah, blah... there was an error...')`?

ryancmoniz commented on server/index.js:
```diff
@@ -1,28 +1,34 @@
+    .catch((err) => {
+      res.status(500).send(err);
+    });
 });
```
Unless you are defining the errors yourself, I would not send back the err object you receive the database. You want to limit the amount of knowledge someone has about your backend as much as possible and database errors give a lot. You could simply send a string back that is short and to the point. "Error getting description from the database"

ryancmoniz commented on client/components/Map.jsx:
```diff
@@ -0,0 +1,30 @@
+import MapGL from 'react-map-gl';
+import PropTypes from 'prop-types';
+
+const MAPBOX_TOKEN = 'pk.eyJ1IjoicnB6aWdsZXIiLCJhIjoiY2tkcXYyeHBhMGpveTJwcGJyYWN3ZnJkMyJ9.konKcHwfSopf4R4nI_fjAQ';
```
Should I be able to see this? Maybe mapbox doesn't have large implications w/r/t a user getting your access token, but we should refrain from making ANY security/authentication related keys viewable on github.

ryancmoniz commented on server/index.js:
```diff
@@ -1,37 +1,41 @@
+  res.sendFile('index.html', { root: path.join(__dirname, '../public') });
+});
+
+app.get('/public/bundle.js', (req, res) => {
+  res.sendFile('bundle.js', { root: path.join(__dirname, '../public') });
+});
+
```
Is there another way we can better serve static assets?

ryancmoniz commented on webpack.config.js:
```diff
@@ -1,32 +1,24 @@
-            },
+        test: /\.m?jsx$/,
+        include: path.resolve(__dirname, 'client'),
+        exclude: /(node_modules|bower_components)/,
```
Are we worried about bower components in this project?

ryancmoniz commented on webpack.config.js:
```diff
@@ -12,7 +12,7 @@ module.exports = {
         test: /\.m?jsx$/,
         include: path.resolve(__dirname, 'client'),
-        exclude: /(node_modules|bower_components)/,
+        exclude: /node_modules/,
```
I see you changed it here! GJ!

ryancmoniz reviewed (commented):
Left a few comments. Remember to not commit your bundle.js file. This will be handled by webpack when neccessary. I saw you also have styled components and prop types in your package.json. Do we intend to use styled components in the future? Are we going to implement prop type checking in all our react files? If not, we can get rid of those. Overall LGTM!

ryanzigler commented:
@ryancmoniz Thanks for the feedback, I really appreciate it!

>You shouldn't directly call lifecycle methods in React. You can modify what they do when something happens (such as getting hostels when the component mounts). But if you feel the need to call it, you should probably modify the way your functions are working. 

I don't remember exactly what I was attempting to do and/or test with that at the time, but it was essentially a quick "fix" to test something. I has since been fixed and lifecycle methods are not being called directly any more.

> Second, if we receive an error, what are we doing? Should we notify the user that something went wrong?

I know that for certain errors, it makes sense to notify the user which is why we see 403 and 404 error pages on websites, but for projects like this where it doesn't make sense to build out an error page, is it a better idea to fail silently or to use something like `alert('blah, blah, blah... there was an error...')`?

> Unless you are defining the errors yourself, I would not send back the err object you receive the database. You want to limit the amount of knowledge someone has about your backend as much as possible and database errors give a lot. You could simply send a string back that is short and to the point. "Error getting description from the database"<br>

👍  Yep, I have already removed that in subsequent commits.  

> Should I be able to see this? Maybe mapbox doesn't have large implications w/r/t a user getting your access token, but we should refrain from making ANY security/authentication related keys viewable on github.

Mapbox actually uses a public key and backend restrictions on that key for security, but good point regardless. I have also since switched to Google Maps instead of Mapbox so I will need to project that API key, which I am currently struggling to do using an .env file as I cannot get the .env file to import into React correctly and all the documentation for that assumes you are using Create React App rather than building everything out yourself. I'll be making a help desk ticket on Monday to help me sort through this issue and schedule office hours after that if needed. But I will make sure not to make any commits with my exposed API key.

> app.get('/public/bundle.js', (req, res) => {
  res.sendFile('bundle.js', { root: path.join(__dirname, '../public') });
});

Yeah, that one wasn't really well thought out. Fixed!
